libraries/chain/smt_objects/nai_pool.cpp

```diff
@@ -0,0 +1,93 @@
+#include <steem/chain/database.hpp>
+#include <steem/protocol/asset_symbol.hpp>
+#include <steem/chain/smt_objects.hpp>
+#include <steem/chain/util/nai_generator.hpp>
+
+#ifdef STEEM_ENABLE_SMT
+
+#define NAI_GENERATION_SEED_BLOCK_ID_HASH_INDEX 4
+
+namespace steem { namespace chain {
+
+/**
+ * Refill the NAI pool with newly generated values
+ *
+ * We will attempt to fill the pool with randomly generated NAIs up until
+ * the pool is full (SMT_MAX_NAI_POOL_COUNT) or we encounter the maximum
+ * acceptable collisions (SMT_MAX_NAI_GENERATION_TRIES). It will once again
+ * replenish when this function is called and the database has a new head
+ * block ID.
+ */
+void replenish_nai_pool( database& db )
+{
+   try
+   {
+      static block_id_type last_block_id = db.head_block_id();
+      static uint32_t collisions_per_block = 0;
+      static uint32_t attempts_per_block = 0;
+
+      // If this is the first time we're encountering this block, reset our variables and set the last block */
+      if ( last_block_id != db.head_block_id() )
+      {
+         last_block_id = db.head_block_id();
```

Comment:
This is minor, but `db.head_block_id()` does a database lookup. While it is likely in cache, it is still common for us to save such values locally, even if they are only used a couple of times.

``` c++
auto head_block_id = db.head_block_id();

if( last_block_id != head_block_id )
{
   last_block_id = head_block_id;
   collisions_per_block = 0;
   attempts_per_block = 0;
}
```